components/builder-db/src/async.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! AsyncServer.register(
+
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread;
+use std::time::Duration;
+use std::fmt;
+
+use rand::{self, Rng};
+use time::SteadyTime;
+use time::Duration as SteadyDuration;
+use threadpool::ThreadPool;
+
+use pool::Pool;
+use error::Result;
+
+pub type DispatchKey = String;
+pub type EventFunction = fn(Pool) -> Result<EventOutcome>;
+
+pub enum EventOutcome {
+    Finished,
+    Retry,
+}
+
+const BACKOFF_SLOT_TIME_MS: u64 = 100;
+const FAILURE_COUNT_UPPER_BOUND: usize = 10;
+
+#[derive(Clone)]
+pub struct AsyncServer {
+    pool: Pool,
+    stop: Arc<AtomicBool>,
+    pub dispatch: Arc<RwLock<HashMap<DispatchKey, EventFunction>>>,
+    pub failure_count: Arc<RwLock<HashMap<DispatchKey, usize>>>,
+    pub retry: Arc<RwLock<HashMap<DispatchKey, (SteadyTime, EventFunction)>>>,
+    pub running: Arc<RwLock<HashSet<DispatchKey>>>,
+}
+
+impl fmt::Debug for AsyncServer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "AsyncServer")
+    }
+}
+
+impl AsyncServer {
+    pub fn new(pool: Pool) -> AsyncServer {
+        AsyncServer {
+            pool: pool,
+            stop: Arc::new(AtomicBool::new(false)),
+            dispatch: Arc::new(RwLock::new(HashMap::new())),
+            failure_count: Arc::new(RwLock::new(HashMap::new())),
+            retry: Arc::new(RwLock::new(HashMap::new())),
+            running: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+
+    pub fn schedule(&self, dispatch_key: &str) -> Result<()> {
+        let has_key = {
+            let d = self.dispatch
+                .read()
+                .expect("Async dispatch lock is poisoned");
+            d.contains_key(dispatch_key)
```

Comment:
I was wanting to both avoid holding two locks at the same time, and give an (admittedly small) chance for a read/write to sneak in from another thread.